### PR TITLE
WIP: add track id to async api calls

### PR DESCRIPTION
**Pull request by DaanHoogland**
## Description

API calls should be logging a more appropriate tracking id. It already exists but seems not be used consistently enough.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: api/src/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java [hunk 1]
```diff
@@ -333,6 +333,7 @@ public boolean getStartVm() {
                     } catch (NumberFormatException e) {
                         throw new InvalidParameterValueException("Unable to translate and find entity with networkId: " + ips.get("networkid"));
                     }
+                    // so no network found but a valid number as network id; are we assuming exceptions thrown or lucky guessing?
```
Comment (rafaelweingartner) [inline api/src/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java]: @DaanHoogland so you are the one that makes the enlightened comments all around ACS code base?! ahahaha

If the network ID is invalid (it was not found), it means that something else further down the deployment process will break, right? If this is the case, I think we could stop the processing here and throw a runtime exception.

Comment (DaanHoogland) [reply api/src/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java]: I don't know yet, maybe an uncaught CloudRuntimeException is thrown somewhere else, but you are right, we can throw it here.

File: server/src/com/cloud/api/ApiServer.java [hunk 2]
```diff
@@ -540,7 +540,7 @@ public String handleRequest(final Map params, final String responseType, final S
                         throw new CloudRuntimeException("No APICommand annotation found for class " + cmdClass.getCanonicalName());
                     }
 
-                    BaseCmd cmdObj = (BaseCmd)cmdClass.newInstance();
+                    BaseCmd cmdObj = (BaseCmd)cmdClass.getConstructor(cmdClass).newInstance();
```
Comment (rhtyd) [inline server/src/com/cloud/api/ApiServer.java]: @DaanHoogland what's the difference in doing `newInstance` directly, than getting the constructor first?

File: server/src/com/cloud/vm/UserVmManagerImpl.java [hunk 3]
```diff
@@ -4600,14 +4602,18 @@ public UserVm createVirtualMachine(DeployVMCmd cmd) throws InsufficientCapacityE
 
         DataCenter zone = _entityMgr.findById(DataCenter.class, zoneId);
         if (zone == null) {
-            throw new InvalidParameterValueException("Unable to find zone by id=" + zoneId);
+            CloudRuntimeException ex =  new InvalidParameterValueException("Unable to find zone by id=" + zoneId);
```
Comment (rhtyd) [inline server/src/com/cloud/vm/UserVmManagerImpl.java]: Does it make sense to instead, get/set the context/jobid in the exception constructor if CallContext is available to it?

File: api/src/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java (hunk 1 above)
Comment (rhtyd) [reply api/src/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java]: Let's remove comments, unless they're around newly added code or explain existing code?

File: server/src/com/cloud/api/ApiServer.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/com/cloud/api/ApiServer.java]: it is deprecated in version 9.

File: api/src/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java (hunk 1 above)
Comment (DaanHoogland) [reply api/src/org/apache/cloudstack/api/command/user/vm/DeployVMCmd.java]: sure, this comment should not add value any more once we decide how to handle this case.
At this point we have not found a network but are having a networkId. I think this should be unreachable.
I will refactor untill the code makes sense to me.

File: server/src/com/cloud/vm/UserVmManagerImpl.java (hunk 3 above)
Comment (DaanHoogland) [reply server/src/com/cloud/vm/UserVmManagerImpl.java]: thanks, yes that's why I created the CloudRuntimeException(String callId, String message) but i haven't created a InvalidParameterValueException(String callId, String message) yet.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: server/src/com/cloud/api/ApiServer.java (hunk 2 above)
Comment (DaanHoogland) [reply server/src/com/cloud/api/ApiServer.java]: ok, stupifd mistake here; i added the class self to the list of parameters :(

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Pull request closed without merging**